File: BackAndServe/domain/audit.py
from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class AuditEntityType(StrEnum):
    USER = "USER"
    CATEGORY = "CATEGORY"
    PRODUCT = "PRODUCT"
    PAYMENT_METHOD = "PAYMENT_METHOD"
    SALE = "SALE"


class AuditActionType(StrEnum):
    INSERT = "INSERT"
    UPDATE = "UPDATE"
    DELETE = "DELETE"
    MIGRATE = "MIGRATE"


class AuditChanges(BaseModel):
    data_before: dict[str, Any] = Field(default_factory=dict)
    data_after: dict[str, Any] = Field(default_factory=dict)


class AuditLogResponse(BaseModel):
    id: int
    username: str | None = None
    description: str | None = None
    action_type: AuditActionType
    entity_type: AuditEntityType
    entity_id: str | None = None
    changes: AuditChanges
    created_at: datetime

File: Front/src/domain/audit.ts
export const AUDIT_ENTITY_TYPES = [
  "USER",
  "CATEGORY",
  "PRODUCT",
  "PAYMENT_METHOD",
  "SALE",
] as const;

export const AUDIT_ACTION_TYPES = [
  "INSERT",
  "UPDATE",
  "DELETE",
  "MIGRATE",
] as const;

export type AuditEntityType = (typeof AUDIT_ENTITY_TYPES)[number];
export type AuditActionType = (typeof AUDIT_ACTION_TYPES)[number];

export type AuditChanges = {
  data_before: Record<string, any>;
  data_after: Record<string, any>;
};

export type AuditLogResponse = {
  id: number;
  username?: string | null;
  description?: string | null;
  action_type: AuditActionType;
  entity_type: AuditEntityType;
  entity_id?: string | null;
  changes: AuditChanges;
  created_at: string;
};

export type AuditListFilters = {
  offset?: number;
  limit?: number;
  start_time?: string;
  end_time?: string;
  username?: string[];
  entity_type?: AuditEntityType[];
  action_type?: AuditActionType[];
};

export type AuditCountFilters = Omit<AuditListFilters, "offset" | "limit">;
